libcudacxx/include/cuda/__mdspan/host_device_accessor.h

```diff
@@ -0,0 +1,273 @@
+ * Host/Device/Managed Accessor Traits
+ **********************************************************************************************************************/
+
+template <typename>
+inline constexpr bool is_host_accessor_v = false;
+
+template <typename>
+inline constexpr bool is_device_accessor_v = false;
+
+template <typename>
+inline constexpr bool is_managed_accessor_v = false;
+
+template <typename _Accessor>
+inline constexpr bool is_host_accessor_v<host_accessor<_Accessor>> = true;
+
+template <typename _Accessor>
+inline constexpr bool is_device_accessor_v<device_accessor<_Accessor>> = true;
+
+template <typename _Accessor>
+inline constexpr bool is_managed_accessor_v<managed_accessor<_Accessor>> = true;
+
+template <typename _Tp>
+inline constexpr bool is_host_device_managed_accessor_v =
+  is_host_accessor_v<_Tp> || is_device_accessor_v<_Tp> || is_managed_accessor_v<_Tp>;
+
+/***********************************************************************************************************************
+ * Host Accessor
+ **********************************************************************************************************************/
+
+template <typename _Accessor>
+struct host_accessor : public _Accessor
+{
+private:
+  using __data_handle_type = typename _Accessor::data_handle_type;
+  using __reference        = typename _Accessor::reference;
+  using __self             = host_accessor<_Accessor>;
+
+  static constexpr bool __is_ctor_noexcept      = noexcept(_Accessor{});
+  static constexpr bool __is_copy_ctor_noexcept = noexcept(_Accessor{});
+  static constexpr bool __is_access_noexcept    = noexcept(_Accessor{}.access(__data_handle_type{}, 0));
+  static constexpr bool __is_offset_noexcept    = noexcept(_Accessor{}.offset(__data_handle_type{}, 0));
+
+  static_assert(!is_host_device_managed_accessor_v<_Accessor>,
+                "cuda::host_accessor/cuda::device_accessor/cuda::managed_accessor cannot be nested");
+
+  template <typename __data_handle_type>
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr bool __is_host_accessible_pointer(__data_handle_type __p) noexcept
+  {
+    if constexpr (_CUDA_VSTD::is_pointer_v<__data_handle_type>)
+    {
+      cudaPointerAttributes __attrib;
+      _CCCL_VERIFY(::cudaPointerGetAttributes(&__attrib, __p) == ::cudaSuccess, "cudaPointerGetAttributes failed");
+      return __attrib.hostPointer != nullptr || __attrib.type == ::cudaMemoryTypeUnregistered;
+    }
+    else
+    {
+      return true; // cannot be verified
+    }
+  }
+
+  _LIBCUDACXX_HIDE_FROM_ABI static constexpr void __check_host_pointer(__data_handle_type __p) noexcept
+  {
+    _CCCL_ASSERT(__self::__is_host_accessible_pointer(__p), "cuda::host_accessor data handle is not a HOST pointer");
+  }
+
+public:
+  using offset_policy = host_accessor;
+
+  _CCCL_HIDE_FROM_ABI host_accessor() noexcept(__is_ctor_noexcept) = default;
```

Comment:
Accessors are not required to be default constructible.  Please see https://eel.is/c++draft/mdspan.accessor.reqmts for the actual requirements, and https://eel.is/c++draft/views.multidim#mdspan.mdspan.cons-1.4 for how mdspan constrains its default constructor on whether its accessor is default constructible.

Reply:
This is pretty easy to fix in C++20.  I have to remind myself how the work-around works in C++ < 20.

```suggestion
  requires(std::is_default_constructible_v<_Accessor>) // needs C++20
  _CCCL_HIDE_FROM_ABI host_accessor() noexcept(__is_ctor_noexcept) = default;
```

Reply:
I tried to fix it. Without c++20, I cannot introduce a template parameter and use `= default`.
Also, I don't know what is the best way to use `requires` before c++20, without introducing a template parameter @miscco